### ISSUE TITLE
I made C64 keyboard controller from Arduino Pro Micro and Soarer firmware for BMC64 use

### ISSUE THREAD

**Issue by mcgurk** [randyrossi/bmc64#6]
This is not an issue, just a comment (is there way to leave comments without raising issue in github?)

I made "poor man's DIY Keyrah" from Arduino Pro Micro (~$3) and Soarer keyboard controller firmware. Commodore key + F7 is F12 and gets you to menu.

https://github.com/mcgurk/Arduino-USB-HID-RetroJoystickAdapter/blob/master/C64_keyboard/README.md

**Comment by randyrossi** [comment]
Very cool!  I will add a link to this project's README in the keyboard section next release.  Thanks!

**Comment by Mantics68** [comment]
Hello, Mcgurk have a .hex for program PRO MICRO? Thank you.

**Comment by mcgurk** [comment]
Soarer_Controller_v1.20_beta4_atmega32u4.hex is for Arduino Pro Micro. It is in Soarer controller firmware package (Soarer_Controller_v1.20_beta4.zip).

**Comment by Mantics68** [comment]
yes thanks I went through links and I saw that there is the firmware in .HEX also for PRO MICRO. Has the problem of symbols on the £ - + * keys been solved with your development?

Inviato da App Alice Mail


lunedì 11 marzo 2019, 09:21 +0100 da notifications@github.com  <notifications@github.com>:

**Comment by mcgurk** [comment]
Problem is that I can't get those five shifted GFX-chars with any keyboard. I don't know what keypresses should be sent to get those buttons shifted. Can you get any how with any keyboard e.g. cross-symbol of plus-button?

**Comment by Mantics68** [comment]
I with the C64USBKEY project I have tried and work all the keys with the modified VKM for WINVICE. if you want, I am attaching them to you and VKM file. good day

Inviato da App Alice Mail


martedì 12 marzo 2019, 09:41 +0100 da notifications@github.com  <notifications@github.com>:

**Comment by mcgurk** [comment]
This 5 key shifting problem is only with BMC64

**Comment by mcgurk** [comment]
When I got "idea" to make C64-keyboard, I thought making my own adapter from Arduino Pro Micro and Arduino IDE. Then I found some readymade Arduino IDE Sketch (could be your github page). Then I remembered that Soarer has made configurable keyboard controller firmware. There is couple of advantages over Arduino IDE. Soarer firmware is not made with Arduino IDE so Arduino Pro Micro shows only as real USB-keyboard and not serial adaptor + keyboard. Some devices doesn't like "double USB devices", so Soarer firmware is more compatible. But maybe most important thing is that it is fully configurable with keyboard matrix pins, macros and etc. without recompiling or reflashing.

**Comment by Mantics68** [comment]
yes I used what I use the sketch drawing from the C64USBKEY program and in fact it is not recognized by the BMC64. at the moment I don't have any other PROMICRO but I ordered them and I will use the file you indicated and maybe I'll try to configure the VKM or in the meantime the developer has solved releasing the new version in.


Inviato da App Alice Mail


martedì 12 marzo 2019, 11:44 +0100 da notifications@github.com  <notifications@github.com>:

**Comment by CemTezcan** [comment]
Amazing because other arduino keyboard approaches doesn't recognized by the BMC64. I made a simple circuit to parallel a C64 keyboard. 

Important thing is you need to add resistors if you connect the c64 keyboard both on c64 motherboard and BMC64 usb input. Otherwise characters are scrambled on key presses. Also video port must be disconnected (it's really strange).

Check my post about that:
https://www.facebook.com/Blockmind/posts/10219813430594876

**Comment by hpingel** [comment]
Too bad that there is no source code available for the Soarer Controller Firmware. I currently have built a C64USBKey (https://github.com/dabonetn/C64USBKey) and it is working on Linux, Windows, etc. but sadly not with BCM64. I would have liked to compare the sources to see what C64USBKey does wrong.

**Comment by mcgurk** [comment]
C64USBKey uses Arduino IDE so it is not pure USB HID-device. Arduino IDE makes Arduino as multiple USB-devices and serial port being first. Simple USB-implementation doesn't see that Arduino device is also keyboard. Aduino IDE libraries could be edited, but better way is just not use Arduino IDE (of Arduino IDE libraries).

**Comment by hpingel** [comment]
Thank you for your explanations. I will try to learn if it is possible to replace the standard bootloader on the Arduino Pro Micro in a way that that there will be no two devices enumerated.

My reason for C64USBKey is simply: It was a project I came across and I built it with the two joystick ports attached (back when I was still playing with Combian). So because it is already soldered and I am lazy I wanted to try to use it with BCM64.

At some point in the future I might build your suggestion when I find time.

**Comment by mcgurk** [comment]
If you do keyboard and joystick adapter with one single Arduino, I think it doesn't work in same reasons. 

With Soarer firmware you can assign all pins freely so no resoldering needed, if you already have working C64USBKey.

Even two joysticks in one Arduino is sometimes problematic. In Windows it works, but Linux detects Arduino Pro Micro as one joystick even when it is two joystick-devices without USB quirks (or changing VID/PID). So Arduino Pro Micro cannot be used as two joystick adapter which works in every situation.

**Comment by hpingel** [comment]
C64USBKey supports 2 joysticks and simply converts the joystick signals to keypresses. The host doesn't know about the joysticks:
https://www.forum64.de/index.php?thread/87674-c64usbkey-c64-usb-keyboard-mit-arduino-pro-micro-oder-teensy-lc/&postID=1373991#post1373991

**Comment by mcgurk** [comment]
Oh I see. You can do same thing easily with Soarer firmware just by modifying config file. No need for recompile or reflashing. Soarer firmware doesn't care if there is keyboard matrix or joystick.

**Comment by mcgurk** [comment]
Here is config file to map keyboard buttons as in your screenshot:
https://github.com/mcgurk/Arduino-USB-HID-RetroJoystickAdapter/blob/master/C64_keyboard/C64_joystick.sc

Your connected pins may be different, so check your pins from here:
https://github.com/mcgurk/Arduino-USB-HID-RetroJoystickAdapter/tree/master/C64_keyboard#pins-i-used

I assumed that GND from joysticks are connected to Arduino GND. Soarer firmware activates internal pullup resistors for input pins and checks when pin goes to GND.

**Comment by K-rnivoro** [comment]
I've made the same using QMK Firmware instead of Soarer's. It's much more powerful, allowing some clever combination, for example Control key is TAB if you tap it, but is actually Control if you hold it. I've also used Teensy++2 so more pins available allowing to connect also the joysticks.
And for the best part, different layouts for "PC" use, than "BMC-64" use, so perfect keyboard emulation.
![2018-06-07 19 04 25](https://user-images.githubusercontent.com/27732136/69596265-ce82f000-0fe0-11ea-8076-933fa247ecda.jpg)

**Comment by hpingel** [comment]
FYI: In the meantime, Dabone has added config files for the soarer firmware to his project C64USBKey: https://github.com/dabonetn/C64USBKey
